flow/layers/picture_raster_cache_item.cc

```diff
@@ -0,0 +1,172 @@
+}
+
+void SkPictureRasterCacheItem::PrerollSetup(PrerollContext* context,
+                                            const SkMatrix& matrix) {
+  cache_state_ = CacheState::kNone;
+  if (!IsPictureWorthRasterizing(sk_picture_, will_change_, is_complex_)) {
+    // We only deal with pictures that are worthy of rasterization.
+    return;
+  }
+  transformation_matrix_ = matrix;
+  transformation_matrix_.preTranslate(offset_.x(), offset_.y());
+  if (!transformation_matrix_.invert(nullptr)) {
+    // The matrix was singular. No point in going further.
+    return;
+  }
+  if (context->raster_cached_entries && context->raster_cache) {
+    context->raster_cached_entries->push_back(this);
+    cache_state_ = CacheState::kCurrent;
+  }
+
+  return;
+}
+
+void SkPictureRasterCacheItem::PrerollFinalize(PrerollContext* context,
+                                               const SkMatrix& matrix) {
+  if (cache_state_ == CacheState::kNone || !context->raster_cache ||
+      !context->raster_cached_entries) {
+    return;
+  }
+  auto* raster_cache = context->raster_cache;
+  SkRect bounds = sk_picture_->cullRect().makeOffset(offset_.x(), offset_.y());
+  // We've marked the cache entry that we would like to cache so it stays
+  // alive, but if the following conditions apply then we need to set our
+  // state back to !cacheable_ so that we don't populate the entry later.
+  if (!SkRect::Intersects(context->cull_rect, bounds)) {
+    cache_state_ = CacheState::kNone;
+    return;
+  }
+
+  // Frame threshold has not yet been reached.
+  if (raster_cache->MarkSeen(key_id_, transformation_matrix_) <
+      context->raster_cache->access_threshold()) {
+    cache_state_ = CacheState::kNone;
+  } else {
+    context->subtree_can_inherit_opacity = true;
+    cache_state_ = CacheState::kCurrent;
+  }
+  return;
+}
+
+bool SkPictureRasterCacheItem::Draw(const PaintContext& context,
+                                    const SkPaint* paint) const {
+  return Draw(context, context.leaf_nodes_canvas, paint);
+}
+
+bool SkPictureRasterCacheItem::Draw(const PaintContext& context,
+                                    SkCanvas* canvas,
+                                    const SkPaint* paint) const {
+  if (!context.raster_cache || !canvas) {
+    return false;
+  }
+  context.raster_cache->Touch(key_id_, canvas->getTotalMatrix());
+  if (cache_state_ == CacheState::kCurrent) {
+    return context.raster_cache->Draw(key_id_, *canvas, paint);
+  }
+  return false;
+}
+
+static const auto* flow_type = "RasterCacheFlow::Picture";
+
+bool SkPictureRasterCacheItem::TryToPrepareRasterCache(
+    const PaintContext& context,
+    bool parent_cached) const {
+  if (!context.raster_cache || parent_cached ||
+      !context.raster_cache->GenerateNewCacheInThisFrame()) {
+    return false;
+  }
+  if (cache_state_ != kNone &&
+      context.raster_cache->MarkSeen(key_id_, transformation_matrix_)) {
```

Comment:
Same comment bout MarkSeen not being needed here as mentioned in the DisplayList variant.